### PR TITLE
Harden batch context container boundaries

### PR DESCRIPTION
**Pull request by Zigfreidish**
## Summary
- Harden retrieval and skill/memory batch prompt-context projection so malformed top-level entry containers fail closed before any iteration.
- Add focused tests and contract documentation for the `entries` container boundary.

## Plan or Spec
- Plan: `docs/plans/2026-06-12-issue-1761-batch-context-container-boundaries.md`
- Spec: `docs/unified-agentic-tool-runtime-contract.md`
- Issue: Refs #1761

## Commands Run
```text
make bootstrap && make proto
# pass

PYTHONPATH="$PWD:$PWD/services/mlx-worker-python" UV_PYTHON=3.12 uv run --project services/mlx-worker-python --extra mlx pytest -q services/mlx-worker-python/tests/test_retrieval_context.py::test_project_retrieval_contexts_refuses_malformed_entry_container services/mlx-worker-python/tests/test_skill_memory_context.py::test_project_skill_memory_contexts_refuses_malformed_entry_container
# red before implementation: 4 failed
# green after implementation: 4 passed in 0.02s

PYTHONPATH="$PWD:$PWD/services/mlx-worker-python" UV_PYTHON=3.12 uv run --project services/mlx-worker-python --extra mlx pytest -q services/mlx-worker-python/tests/test_retrieval_context.py services/mlx-worker-python/tests/test_skill_memory_context.py services/mlx-worker-python/tests/test_prompt_context.py
# 63 passed in 0.06s

PYTHONPATH="$PWD:$PWD/services/mlx-worker-python" UV_PYTHON=3.12 uv run --project services/mlx-worker-python --extra mlx coverage run -m pytest -q services/mlx-worker-python/tests/test_retrieval_context.py services/mlx-worker-python/tests/test_skill_memory_context.py services/mlx-worker-python/tests/test_prompt_context.py
PYTHONPATH="$PWD:$PWD/services/mlx-worker-python" UV_PYTHON=3.12 uv run --project services/mlx-worker-python --extra mlx coverage json -o coverage.json
UV_PYTHON=3.12 uv run --project services/mlx-worker-python --extra mlx python scripts/changed_scope_coverage.py --coverage-json coverage.json services/mlx-worker-python/worker/runtime/retrieval_context.py services/mlx-worker-python/worker/runtime/skill_memory_context.py services/mlx-worker-python/tests/test_retrieval_context.py services/mlx-worker-python/tests/test_skill_memory_context.py
# TOTAL 7 0 100%

PYTHONPATH="$PWD:$PWD/services/mlx-worker-python" UV_PYTHON=3.12 uv run --project services/mlx-worker-python --extra mlx coverage run -m pytest -q services/mlx-worker-python/tests/test_retrieval_context.py::test_project_retrieval_contexts_admits_multiple_entries_with_redacted_receipts services/mlx-worker-python/tests/test_retrieval_context.py::test_project_retrieval_contexts_isolates_refusals_without_dropping_valid_entries services/mlx-worker-python/tests/test_retrieval_context.py::test_project_retrieval_contexts_refuses_malformed_entry_objects_without_dropping_valid_entries services/mlx-worker-python/tests/test_retrieval_context.py::test_project_retrieval_contexts_refuses_duplicate_payload_fields_before_overwrite services/mlx-worker-python/tests/test_retrieval_context.py::test_project_retrieval_contexts_refuses_duplicate_with_defensive_receipt_fallbacks services/mlx-worker-python/tests/test_pr_scoped_performance.py::test_scope_report_selects_retrieval_context_projection_probe services/mlx-worker-python/tests/test_pr_scoped_performance.py::test_retrieval_context_projection_probe_script_emits_metrics services/mlx-worker-python/tests/test_pr_scoped_performance.py::test_registered_probes_expose_focused_commands services/mlx-worker-python/tests/test_pr_scoped_performance.py::test_registered_probe_registry_entries_validate_commands_and_watch_globs
PYTHONPATH="$PWD:$PWD/services/mlx-worker-python" UV_PYTHON=3.12 uv run --project services/mlx-worker-python --extra mlx coverage json -o coverage.json
UV_PYTHON=3.12 uv run --project services/mlx-worker-python --extra mlx python scripts/changed_scope_coverage.py --coverage-json coverage.json services/mlx-worker-python/worker/runtime/retrieval_context.py services/mlx-worker-python/tests/test_retrieval_context.py services/mlx-worker-python/tests/test_pr_scoped_performance.py scripts/retrieval_context_projection_probe.py
# 9 passed in 0.13s; TOTAL 7 0 100%

git diff --check
# pass

git commit --amend --no-edit
# pre-commit pass before origin/main merge:
# make swift-test: pass, elapsed 141.8s
# make py-test: 3944 passed, 14 skipped, 2 warnings in 162.29s
# make integration-test: 120 passed, 1 skipped in 414.58s
# scoped performance: Status ok, selected probes 1, regressions 0, verification failures 0

git merge origin/main --no-edit
# pass, no conflicts; origin/main advanced to 0e4b57fdaecee601f4ff1b3248a36ec817d79e7b

PYTHONPATH="$PWD:$PWD/services/mlx-worker-python" UV_PYTHON=3.12 uv run --project services/mlx-worker-python --extra mlx coverage run -m pytest -q services/mlx-worker-python/tests/test_retrieval_context.py services/mlx-worker-python/tests/test_skill_memory_context.py services/mlx-worker-python/tests/test_prompt_context.py
PYTHONPATH="$PWD:$PWD/services/mlx-worker-python" UV_PYTHON=3.12 uv run --project services/mlx-worker-python --extra mlx coverage json -o coverage.json
UV_PYTHON=3.12 uv run --project services/mlx-worker-python --extra mlx python scripts/changed_scope_coverage.py --coverage-json coverage.json services/mlx-worker-python/worker/runtime/retrieval_context.py services/mlx-worker-python/worker/runtime/skill_memory_context.py services/mlx-worker-python/tests/test_retrieval_context.py services/mlx-worker-python/tests/test_skill_memory_context.py
# 63 passed in 0.13s; TOTAL 0 0 100% after merge commit comparison

make swift-test
# pass

make py-test
# 3947 passed, 14 skipped, 2 warnings in 164.40s

make integration-test
# 120 passed, 1 skipped in 677.31s

PYTHONPATH="$PWD:$PWD/services/mlx-worker-python" UV_PYTHON=3.12 uv run --project services/mlx-worker-python --extra mlx python scripts/pre_commit_gate.py
# N/A: hook-style run has no staged files after merge commit; scoped performance was run manually with run_performance_report(..., base_ref="origin/main") instead.

PYTHONPATH="$PWD:$PWD/services/mlx-worker-python" UV_PYTHON=3.12 uv run --project services/mlx-worker-python --extra mlx python -c 'from pathlib import Path; from scripts.pre_commit_gate import run_git, run_performance_report; changed=[line.strip() for line in run_git(Path.cwd(), ["diff", "--name-only", "origin/main...HEAD"]).splitlines() if line.strip()]; outcome=run_performance_report(Path.cwd(), changed, base_ref="origin/main"); raise SystemExit(0 if outcome.status == "ok" else 1)'
# scoped performance: Status ok, selected probes 2, regressions 0, context regressions 0, verification failures 0
```

## Coverage and Metrics
- Changed-scope coverage: `TOTAL 7 0 100%` for the changed retrieval and skill/memory runtime/test scope.
- Pre-commit scoped performance report before origin/main merge: `.runtime/pre-commit-performance/20260612-035851-ccbe12ed/report/report.md`
- Post-merge local scoped performance report: `.runtime/pre-commit-performance/20260612-050952-baedf355/report/report.md`
- Local report summary after merging latest `origin/main`: `Status: ok`, selected probes `2`, regressions `0`, context regressions `0`, verification failures `0`.
- Remote PR scoped performance comment before the post-merge push: `Status: ok`, selected probes `2`, regressions `0`, context regressions `0`, verification failures `0`.
- Retrieval context projection fast path: targeted tests pass, coverage pass `100.0%`; local `optimized_elapsed_ms_mean` improved from `0.028ms` to `0.018ms`, and remote `optimized_elapsed_ms_mean` improved from `0.091ms` to `0.046ms`.
- Observability mode: `evidence`; this change adds no new runtime metrics or debug artifacts. Probe overhead is bounded to existing PR-scoped performance probes.

## Known Gaps
- This slice does not implement new retrieval, skill, or memory stores.
- This slice does not add owner-scope lookup enforcement beyond preserving existing per-entry receipt fields.
- This slice only hardens shared helper top-level batch containers; remaining #1761 entrypoint/store boundaries are deferred to later slices.
- Gemini review suggested narrowing the raw-boundary parameter types and using `isinstance`; replied and resolved because `Any` plus exact list/tuple checks are intentional fail-closed untrusted-boundary behavior.

## Evidence Checklist
- [x] Relevant plan or spec is identified.
- [x] Behavior changes are reflected in the relevant docs.
- [x] Protocol changes include regenerated generated artifacts. N/A: no protocol changes.
- [x] Dependency changes include updated lockfiles. N/A: no dependency changes.
- [x] Relevant tests were run.
- [x] A metrics report is included, or `N/A` is stated explicitly with the reason.
- [x] Observability mode, probe overhead, and any deferred debug or evidence work are recorded, or `N/A` is stated explicitly with the reason.
- [x] Deferred work and known gaps are stated explicitly.

### CODE REVIEW

File: services/mlx-worker-python/worker/runtime/skill_memory_context.py [hunk 1]
```diff
@@ -88,8 +88,24 @@ def admit_memory_context(
 
 
 def project_skill_memory_contexts(
-    entries: list[SkillMemoryContextEntry] | tuple[SkillMemoryContextEntry, ...],
+    entries: Any,
 ) -> SkillMemoryContextProjection:
+    entries_type = type(entries)
+    if entries_type is not list and entries_type is not tuple:
```
[the remark this answers is not in the document]
Comment (Zigfreidish) [reply services/mlx-worker-python/worker/runtime/skill_memory_context.py:94]: For this boundary, keeping `Any` is intentional: this helper accepts raw prompt-context input and must convert malformed top-level values into typed refusal receipts instead of assuming the caller already has a typed container. The exact `type(entries) is list/tuple` check is also intentional because list and tuple are the only admitted top-level containers; `isinstance` would widen this untrusted boundary to custom container subclasses. The added tests cover dict/string containers with `source_field="entries"`, and the latest scoped performance report is `ok` with 0 regressions. No code change here.

File: services/mlx-worker-python/worker/runtime/retrieval_context.py [hunk 2]
```diff
@@ -88,8 +88,23 @@ def admit_retrieved_image_context(
 
 
 def project_retrieval_contexts(
-    entries: list[RetrievalContextEntry] | tuple[RetrievalContextEntry, ...],
+    entries: Any,
 ) -> RetrievalContextProjection:
+    entries_type = type(entries)
+    if entries_type is not list and entries_type is not tuple:
```
[the remark this answers is not in the document]
Comment (Zigfreidish) [reply services/mlx-worker-python/worker/runtime/retrieval_context.py:94]: For this boundary, keeping `Any` is intentional: this helper accepts raw prompt-context input and must convert malformed top-level values into typed refusal receipts instead of assuming the caller already has a typed container. The exact `type(entries) is list/tuple` check is also intentional because list and tuple are the only admitted top-level containers; `isinstance` would widen this untrusted boundary to custom container subclasses. The added tests cover dict/string containers with `source_field="entries"`, and the latest scoped performance report is `ok` with 0 regressions. No code change here.

File: services/mlx-worker-python/tests/test_retrieval_context.py [hunk 3]
```diff
@@ -575,6 +601,9 @@ def test_project_retrieval_contexts_isolates_refusals_without_dropping_valid_ent
 
 
 def test_project_retrieval_contexts_refuses_malformed_entry_objects_without_dropping_valid_entries() -> None:
```
Comment (Zigfreidish) [inline services/mlx-worker-python/tests/test_retrieval_context.py:603]: Minor style nit (non-blocking): the skill/memory side gets a dedicated `@pytest.mark.parametrize` test (`test_project_skill_memory_contexts_refuses_malformed_entry_container`) for the container-level refusal scenario, which makes each input a first-class named test case. Here the same scenario is embedded as two helper calls at the top of the existing `_refuses_malformed_entry_objects_without_dropping_valid_entries` test, which bundles two distinct refusal levels (container vs. per-entry) into one test function.

Consider extracting a dedicated parametrized test for the container case to match the skill/memory pattern and keep per-entry and container failures independently visible in the test report. No behavior change needed — purely a readability/consistency suggestion for a future cleanup.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by Zigfreidish — COMMENTED** [review]
## Review — Harden batch context container boundaries

**Overview**

Adds a fail-closed top-level container type guard to both `project_retrieval_contexts` and `project_skill_memory_contexts`. When the `entries` argument is not a `list` or `tuple`, the function returns immediately with no user payload, no admitted receipts, and one typed refusal receipt (`source_field = "entries"`), rather than iterating arbitrary values or leaking Python exceptions.

**Code correctness** ✅

The exact-type check (`type(entries) is not list and type(entries) is not tuple`) is intentionally stricter than `isinstance` — it refuses subclasses, which is the right posture at an untrusted boundary. This is consistent with the project's fail-closed boundary philosophy and is documented in the PR's Known Gaps.

Widening the parameter annotation to `Any` is correct: the guard itself is the runtime enforcement, and the old typed union would reject the guard body under a strict checker. The public contract is now documented in `docs/unified-agentic-tool-runtime-contract.md`.

**Bundled micro-optimization** ✅

The `receipts_extend(generator)` → `for` loop + `receipts_append` change in `retrieval_context.py` is a consistent application of the existing method-reference caching pattern in that file. Trivially correct and narrowly scoped.

**Test coverage** ✅

- Four new test cases: dict and string containers for both retrieval and skill/memory helpers.
- Skill/memory uses `@pytest.mark.parametrize`; retrieval uses an inline helper — see the inline comment for a minor future style suggestion.
- Red-before-green confirmed in the PR description.
- Changed-scope coverage: `TOTAL 7 0 100%`.
- Pre-commit gate: swift-test pass, 3944 py-tests pass, 120 integration tests pass, scoped-performance `ok`, regressions `0`.

**Security** ✅

The change strengthens the safety posture: a malformed batch container no longer silently iterates arbitrary content into the prompt. Fail-closed is the correct default here.

One minor inline comment left on the retrieval test for a future style cleanup. No blocking concerns. LGTM — would approve if GitHub permitted self-review.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: services/mlx-worker-python/tests/test_retrieval_context.py (hunk 3 above)
Comment (Zigfreidish) [reply services/mlx-worker-python/tests/test_retrieval_context.py:603]: Leaving this as a follow-up rather than changing this slice now. The retrieval container assertions are deliberately folded into the existing retrieval probe-covered test so the changed boundary lines stay covered by the current retrieval performance probe without expanding the probe registry and forcing unrelated probes. The skill/memory side has a dedicated parametrized test because it is not tied to that performance probe coverage path. No behavior change here.
